Added initial draft for filtering design proposal [WIP]

rhuss opened the pull request.

rhuss committed:
Added initial draft for filtering proposal [WIP]

jimmidyson commented on design-proposals/002-filter-data.md:
```diff
@@ -0,0 +1,41 @@
+* Based on content in the body.
+* Based on content in a message header.
+
+In both cases, I expect to be able to apply basic string operations to the data to form a filter expression (contains, does not contain, starts with).  For more details on options, see the Operator Support section here:
```
> contains, does not contain, starts with

Is that just illustrative of what is possible? I assume we want to support the full set of simple language filters at http://camel.apache.org/simple.html#Simple-OperatorSupport?

rhuss replied:
its just illustrative (should have added ellipsis at the end)

rhuss replied:
It would be awesome though when we could restrict the set of operators based on the data-type of the field selected.

jimmidyson replied:
Nit: add `etc` here to indicate these are examples and not an exhaustive list.

jimmidyson commented on design-proposals/002-filter-data.md:
```diff
@@ -0,0 +1,41 @@
+In both cases, I expect to be able to apply basic string operations to the data to form a filter expression (contains, does not contain, starts with).  For more details on options, see the Operator Support section here:
+http://camel.apache.org/simple.html
+
+Similar to logging, it would be nice if the UI provided help in creating filter expressions for new/inexperienced users (like through clickable examples).  The interface should also allow for advanced users to enter free form expressions.
```
> it would be nice

I think we should be more prescriptive here. Does this proposal set out that this should happen (I think it should personally)? If so, then this should be stated positively.

rhuss replied:
agreed with both.

jimmidyson commented on design-proposals/002-filter-data.md:
```diff
@@ -0,0 +1,41 @@
+
+#### syndesis-ui
+
+* Evalute which editing modes should be supported
```
Typo: `evaluate`

jimmidyson reviewed (changes requested):
IMO this reads more like an issue for discussion than a firm proposal. I feel that proposals should be more prescriptive, but of course we need to allow for changes once we get to implementation.

rhuss committed:
Update according to review

rhuss commented:
@jimmidyson as stated its just a WIP. _It's not even alpha_ Actually its more like a brainstorming skeleton which I will fill out. I only opened the PR so soon to check the ZenHub process.

rhuss committed:
Update

rhuss committed:
Fix JSON

rhuss committed:
Formatting

rhuss commented:
@jimmidyson added some more substance to this proposal so that the implementation on the server side should be straight forward. Would be awesome if you could have a review on this.

rhuss committed:
Update to reflect the fact that simple does not support operator precedence.

jimmidyson commented on design-proposals/002-filter-data.md:
```diff
@@ -0,0 +1,142 @@
+  value: "antman"
+- type: "header"    
+  key: "region"
+  op: "=~"
```
I know that camel combines both symbols and words (e.g. `==` and `regex`) for ops but should we just copy that or use descriptive words and map appropriately at runtime?

rhuss replied:
I think we need some sort of meta data anyway for the list of available operators. This meta data could contain both, a key, a description and the operator used in simple. In that case we would store the key here.

My biggest concern wrt meta data is how to get the possible header keys to select from. Any ideas ?

jimmidyson replied:
Not right now :)

Back to the original question, are we going to stick with the same ops as camel simple (combination of symbols and words) or use descriptive words consistently?

rhuss replied:
The UI needs to fetch the list of possible operators anyway somehow (would not hardcode it in the UI but let the UI be agnostic about the possible operators so that the list of the operators should be returned by syndesis-rest). So we can make it richer to set the operator within the object and use the operator-description mapping from the server.

jimmidyson commented on design-proposals/002-filter-data.md:
```diff
@@ -0,0 +1,142 @@
+* **type** : On which part of the message to apply the filter. Can be either "header", "body" or "properties".
+* **key** : The key to use to extract a message value (e.g. header or property name). The key is empty when type is equals to body.
+* **op** : The operator to use for the filter
+* **value** : The value used by the operator to determine whether a filter applies.
```
Do we care about unary operators?

davsclaus replied:
No we should not

jimmidyson commented on design-proposals/002-filter-data.md:
````diff
@@ -0,0 +1,142 @@
+${body} contains "antman" && ${in.header.region} =~ "asia" || ${body} regex "bat(wo)?man" && ${in.header.publisher} =~ "DC Comics"
+```
+
+The simple expression language [does not support parentheses](http://camel.apache.org/simple.html) nor precedence of operators so the expression is always evaluated from left to right.
````
@davsclaus Is there any plan to support parentheses? It limits what we can easily represent in the filter.

davsclaus replied:
No the simple language works well today for normal use-cases.

rhuss replied:
The alternative would be to switch the expression language to e.g. groovy.

jimmidyson replied:
The good thing about simple is that it is, um, simple and restricted. I do expect to support groovy as a scripting step, but not sure if we want to expand the filter step right now to be able to use anything other than simple so we can get a first cut of this working.

Now we're talking about it though, should we add the predicate language to the filter step, even if that is just simple for now?

rhuss replied:
Good question. From an UI POV it would be simpler to have only a single expression. For an expanded use case one could then go to the text field. I wouldn't mind to omit 'and' and 'or' for now.

jimmidyson commented on design-proposals/002-filter-data.md:
```diff
@@ -0,0 +1,142 @@
+  "stepKind": "filter",
+  "configuredProperties": {
+    "type": "form",
+    "simple" : "${body} contains \"antman\" || ${in.header.publisher} =~ \"DC Comics\"",
```
Why do we need to store the simple expression here along with the rules?

rhuss replied:
I think it's easier for the generation of the actual step for the runtime function.yaml to no replicate the logic to build up the simple expression (which we need when we storing the rule to validate the expression). Also, the runtime generator then doesn't care whether the expression has been created as free text or via a form and just picks the content of this field.

The price is that we add some denormalization and the possibility for some inconsistencies.

jimmidyson replied:
Sounds good.

jimmidyson commented on design-proposals/002-filter-data.md:
````diff
@@ -0,0 +1,142 @@
+}]
+```
+
+The example is simplified in so far as the value to the "rules" field for step with id "1" must be enterd as a single line string with newlines replaced by "\n" since the value of a property is currently only allowed to be a string. 
````
nit: typo `entered`

rhuss replied:
fixed

jimmidyson commented on design-proposals/002-filter-data.md:
```diff
@@ -0,0 +1,142 @@
+
+### Remarks
+
+* The current domain object `IntegrationConnectionStep` seems to be supeflous and should be deleted then.
```
nit: typo `superfluous`

rhuss replied:
fixed

rhuss committed:
Adressed review comments

jimmidyson reviewed (commented):
A few outstanding questions - let's try to get this merged ASAP though so quick response would be good (I'll endeavour to do the same!).

rhuss commented:
sorry, for the late response. get distracted by meetings (woohoo...) and then oversaw this update as it was on the second page of my octobox.

rhuss committed:
Added section on how to obtain background data.

rhuss commented:
@jimmidyson I added a section how to get the background data for possible selection on keys and operators.

jimmidyson commented on design-proposals/002-filter-data.md:
```diff
@@ -67,7 +67,30 @@ would be stored as a property "filter" on the integration step. It translates la
+
+For the `keys` as well as for the possible `operators` the UI needs a list of values which can be chosen. The list of operators should be fixed, whereas it should be possible to add a freeform key (but with suggestion of a set of given keys). 
+
+This background data can be obtained by a dedicated API call to an endpoint `../integrations/filter/options` which takes an existing integration ID (if the intergration has already been created, otherwise the ids of all connections before this filter step need to be send to the API server) as parameter.
```
Do we need two separate endpoints? How about just sending through the connection IDs to keep it consistent between existing integrations and new integrations?

rhuss replied:
I thought about caching the "operators" list as this is supposed not to change during a session. But we could send it all the time, too along with the connection specific keys, too.

kahboom commented:
@davsclaus @rhuss @jimmidyson - For the "contains" filter it's not clear whether or not this means "contains exact phrase" and if that is case-sensitive. Would you mind clarifying? This should be clear to the user what their filter will do exactly.

For instance, when Googling you would use quotation marks for exact phrases, and I think it would be case insensitive (e.g. "blue airplane", "BLUE airplane" would yield the same results). Is this possible on the backend? Also, if not, then the phrase blue airplane without phrases should be treated as two separate keywords in Google, where "blue airplane" takes priority, but if results are low then "blue" results and "airplane" results are also displayed to the user.

If that's the case, our dropdown check should be "contains exact" not just "contains". If both are acceptable, then we'd need to let the user know that as well.

In short:
1. What is accepted/possible on the backend?
2. What do we want to be possible for the user?

cc @gashcrumb @dongniwang

davsclaus commented:
Beware that the Camel operator `=~` is equals ignore case, so they two texts must match.

And the Camel `contains` operator is using the Java String `contains` method which is case sensitive.

There is no operator for ignore case for contains operator.

rhuss commented:
@kahboom In fact there should be "contains", "equals" and "equals (ignore case)" as options (which then map to the operators @davsclaus mentioned).

rhuss commented:
@davsclaus I guess one could simulate "contains (ignore case)" with a regexp operator ?

davsclaus commented:
Yeah you can do almost anything with regexp, but then 2 months later fell the pain if the reg exp got too complex.

I wonder if we can add a native contains (ignore case) operator to Camel. Because contains is not using a symbol it may be harder to come up with a nice short name. Or we could think about using the `~` symbol here also - `~contains` or spell it out `contains-ignore-case`.

I am not sure if you want to make it ignore case by default, eg contains. How often would you have the need for case sensitive?

rhuss commented:
For the sake of Syndesis we could hide the complexity of a regexp behind the "contains (ignore case)". This is intended for the citizen user to be used via a drop down box and we then could add the translation to a regexp also to the mapping which we have in order to map the simple operator to UI names. 

Having this natively, though would be nice of course, too. Whats about to have e.g. '~~' for matching to 'contains (ignore case)' ? I wouldn't change the current semantics of 'contains' as this would be not backwards compatible (and you never know if e.g. people try to find out names like 'Apple' in contrast to the fruit 'apple').

dongniwang commented:
Hello everyone, here's the updated designs for filter based on feedback from our discussion on Monday. https://redhat.invisionapp.com/share/9BC79TYHZ

Please review designs and let me know if you have any feedback! Happy to meet to discuss further if needed! Thank you!

davsclaus commented:
Added a ticket about adding the contains ignore case operator to Camel
https://issues.apache.org/jira/projects/CAMEL/issues/CAMEL-11420

rhuss commented:
Thanks a lot for the comments. I will try to update the proper links and align the API descriptions with the format in @zregvart credential proposal as it nicer than giving JSON examples only.

kahboom commented:
Thanks @davsclaus that would be really helpful, as users would be very annoyed if they had to create a separate integration for matching "SYNDESIS", "syndesis", "Syndesis", etc. and even then it wouldn't capture all possible tweets, BUT like @rhuss says, having case-sensitive could be useful for things like Apple vs apple (although, I do question whether or not they should just be using the @ mention feature or contains "@apple" or "#apple"). The user won't be able to control how people tweet though, so someone that wants more fine-grained control could definitely want that feature, but as a final consideration, I think this same user should instead be using "EXACT" phrases like that might require quotation marks (need to see the Camel docs sorry), but that would be case sensitive as it is with Google search. In general I think we should be following Google search rules. 😂

davsclaus commented:
Yeah so the CAMEL-11420 has been implemented and the operator is `~~` as suggested by Roland. This will be in the Camel 2.20.0 release.

rhuss committed:
Update spec

rhuss commented:
@jimmidyson I think I addressed all your concerns. I hope its ok when I merge now while you are still on PTO. We can still continue to work on it but I think its good enough to start with the implementation.
